### DISCUSSION TITLE
Allow admins to control the password strength

### DISCUSSION THREAD

**Discussion by MajliTech** [Mailu/Mailu#3074] [Feature requests / Ideas]
Hello! I have just moved to a new server which is Mailu-powered. However, I am targeting spammers wanting to control my server. So, to combat that, I need a way to enforce some password rules.

I believe this could be implemented using Reg Ex, for example

- User friendly name: at least 8 letters
regex: [a-zA-Z]{8}
must match: at least 1
- User friendly name: 1 digit
regex: [0-9]
must match: at least 1 times
- User friendly name: cannot contain your username, your email or your server domain
regex: {USERNAME}|{EMAIL}|{DOMAIN}
must match: exactly 0 times
...
and so on. 

The server admin could define their own regex rules, so the power is still in their hands. I wanted to write that feature, I'm just too dumb for that now. Any feedback is appreciated.

**Comment by nextgens** [comment]
Mailu has some very powerful rate-limiting capabilities; Attackers are very unlikely to be able to bruteforce passwords.

NIST's view is that "[No other complexity requirements for memorized secrets SHOULD be imposed](https://pages.nist.gov/800-63-3/sp800-63b.html#5111-memorized-secret-authenticators)"

In the upcoming release you have ``AUTH_REQUIRE_TOKENS`` (see https://mailu.io/master/configuration.html#advanced-settings) that will alleviate the problem altogether.

**Comment by MajliTech** [comment]
I already had successful account takeovers. When I get some spare time, I'll code it up.

Nov 28, 2023, 10:26 PM by ***@***.***:

**Reply by nextgens** [reply]
Account takeover when running a recent version of Mailu?

Was it credential stuffing and the user reusing his password or something else?

What did you have the rate-limits set to?

**Comment by MajliTech** [comment]
Of course password reuse with a very simple password, that's why I suggest this. While you can have client tokens, signing into Maili can be still does with a user defined password like football1995. Of course I am overexaggerating, but some other weak passwords can still be used.

Nov 29, 2023, 8:27 AM by ***@***.***:

**Reply by nextgens** [reply]
That's the thing: if it is password reuse then "complexity" won't help. The attacker may replay as easily a complex password than a simple one.

**Reply by MajliTech** [reply]
I have one last argument: if the database leaked (weak ssh password, smh) then the hash cracking would take longer.

**Reply by nextgens** [reply]
That is a valid but not good argument. We use bcrypt (a very slow hash function) and you can configure the ``CREDENTIAL_ROUNDS`` value higher if that is something you want to protect against.

... ultimately your users should use application passwords/tokens for their thick clients and not their passwords.

**Comment by MajliTech** [comment]
So, what you saying is, the password requirements are going to give users a false sense of security?

Nov 29, 2023, 12:09 PM by ***@***.***:

**Reply by nextgens** [reply]
I invite you to read https://pages.nist.gov/800-63-3/sp800-63b.html#a3-complexity
